### PR TITLE
[inductor] Error on unsupported autotuner configs

### PR DESCRIPTION
**Pull request by SamGinzburg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139658

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by SamGinzburg** [comment]
Addresses https://github.com/pytorch/pytorch/issues/139059 by erroring on pre/post hook in autotuner config & Config

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 1]
```diff
@@ -1068,6 +1067,21 @@ def init_variable(
...
+                    )
+                    # we also cannot support pre/post hook in the autotuner config
+                    or (
+                        "pre_hook" in defaults
+                        and defaults["pre_hook"] != kernel.pre_hook
+                    )
+                    or (
+                        "post_hook" in defaults
+                        and defaults["post_hook"] != kernel.post_hook
+                    )
```
Comment (aakhundov) [inline torch/_higher_order_ops/triton_kernel_wrap.py:1084]: Seems the `pre_hook` and `post_hook` attributes of `Autotuner` are lambdas by default: https://github.com/triton-lang/triton/blob/f484cb8f00e4222c7db81b47c58e88467fe7a56b/python/triton/runtime/autotuner.py#L55-L56 . So these conditions will always be `True`. I don't see an obvious way to check if the `pre_hook` / `post_hook` have been passed (and are not defaults), as we don't intercept the `@triton.autotune` call in Dynamo, just the object that is used in the PT2-compiled code.

Perhaps, let's limit this PR to configs now and add flags to upstream Triton to mark that the `Autotuner`'s pre_/post_hook have been set? Then we can look at those flags here to error out.

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 2]
```diff
@@ -1068,6 +1067,21 @@ def init_variable(
                         "use_cuda_graph" in defaults
                         and defaults["use_cuda_graph"].default != kernel.use_cuda_graph
                     )
+                    # pre_hook requires running arbitrary code at runtime, which we cannot handle at this time
+                    # https://github.com/pytorch/pytorch/issues/139059
+                    or (
+                        # Check Config passed to autotuner in configs
+                        any(cfg.pre_hook is not None for cfg in kernel.configs)
```
Comment (aakhundov) [inline torch/_higher_order_ops/triton_kernel_wrap.py:1074]: Should we check this separately to make the error message more specific?

Comment (SamGinzburg) [reply torch/_higher_order_ops/triton_kernel_wrap.py:1074]: yeah the only configs & keys one is a bit vague. maybe something like "pre_hook is not supported in triton.Autotune Configs"?

File: test/inductor/test_triton_kernels.py [hunk 3]
```diff
@@ -3377,6 +3377,61 @@ def grid(META):
             "grid_wrapper_for_op_zeros_0"
         ).check_next("return (256").check_next("return (64").run(output)
 
+    @requires_gpu
+    def test_autotune_no_pre_or_post_hook(self):
+        import triton
+        import triton.language as tl
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:3383]: Do we need to re-import this? It should be imported at the top.

File: test/inductor/test_triton_kernels.py [hunk 4]
```diff
@@ -3377,6 +3377,61 @@ def grid(META):
...
+            n_elements = output.numel()
+            grid = lambda meta: (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
+            add_kernel[grid](x, y, output, n_elements)
+            return output
+
+        x = torch.ones((4096,), device="cuda:0", dtype=torch.float16)
+        y = torch.ones((4096,), device="cuda:0", dtype=torch.float16)
+
+        # should always pass
+        assert add(x, y).mean() == 2, "Problem with add kernel"
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:3427]: Nit: let's stay consistent with other tests (smaller random inputs, direct output comparison, `GPU_TYPE`)? [Example](https://github.com/pytorch/pytorch/blob/639162f39a8a9b17fced61795dfcd93cdcb1c044/test/inductor/test_triton_kernels.py#L3246).

File: test/inductor/test_triton_kernels.py [hunk 5]
```diff
@@ -3377,6 +3377,61 @@ def grid(META):
...
+
+        # pre_hook requires running arbitrary code at runtime, which we cannot handle at this time
+        # https://github.com/pytorch/pytorch/issues/139059
+        def get_default_config():
+            config = triton.Config(
+                {"BLOCK_SIZE": 1024},
+                num_warps=4,
+                num_stages=2,
+                pre_hook=init_to_zero("output_ptr"),
+            )
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:3396]: Nit: can we inline this into `@triton.autotune(configs=...)` for simplicity?

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 6]
```diff
@@ -1073,6 +1072,16 @@ def init_variable(
                 self.raise_unsupported(
                     "Only configs and keys are supported for triton.autotune"
                 )
+            if not (
+                torch._inductor.config.unsafe_ignore_unsupported_triton_autotune_args
+                # pre_hook requires running arbitrary code at runtime, which we cannot handle at this time
+                # https://github.com/pytorch/pytorch/issues/139059
+                # Check Config passed to autotuner in configs
+                and (any(cfg.pre_hook is not None for cfg in kernel.configs))
+            ):
```
[the remark this answers is not in the document]
Comment (aakhundov) [reply torch/_higher_order_ops/triton_kernel_wrap.py:1081]: I'm actually surprised the other tests are passing.

**Review by aakhundov — APPROVED** [review]

**Pull request closed without merging**